contributors/design-proposals/kubelet-rootfs-distribution.md

```diff
@@ -0,0 +1,191 @@
+
+
+#### Is this running the kubelet as a pod?
+
+This is different than running the Kubelet as a pod. Rather than using namespaces, it uses only a chroot and shared bind mounts.
+
+## Alternatives
+
+#### Container + Shared bindmounts
+
+Instead of using a chroot with shared bindmounts, a proper pod or container could be used if the container supported shared bindmounts.
+
+This introduces some additional complexity in requiring something more than just the bare minimum. It also relies on having a container runtime available and puts said runtime in the critical path for the Kubelet.
+
+#### "Dependency rootfs" aware kubelet
+
+The Kubelet could be made aware of the rootfs containing all its dependencies, but not chrooted into it (e.g. started with a `--dependency-root-dir=/path/to/extracted/container` flag).
+
+The Kubelet could then always search for the binary it wishes to run in that path first and prefer it, as well as preferring libraries in that path. It would effectively run all dependencies similar to the following:
+
+```bash
+export PATH=${dep_root}/bin:${dep_root}/usr/bin:...
+export LD_LIBRARY_PATH=${dep_root}/lib:${dep_root}/usr/lib:...
+# Run 'mount':
+$ ${dep_root}/lib/x86_64-linux-gnu/ld.so --inhibit-cache mount $args
+```
+
+**Downsides**:
+
+This adds significant complexity and, due to the dynamic library hackery, might require some container-specific knowledge of the Kubelet or a rootfs of a predetermined form.
+
+This solution would also have to still solve the packaging of that rootfs, though the solution would likely be identical to the solution for distributing the chroot-kubelet-rootfs.
+
+#### Waiting for Flexv2 + port-forwarding changes
+
+The CRI effort plans to change how [port-forward](https://github.com/kubernetes/kubernetes/issues/29579) works, towards a method which will not depend explicitly on socat or other networking utilities.
+
+Similarly, for the mount utilities, the [Flex Volume v2](https://github.com/kubernetes/features/issues/93) feature is aiming to solve this utility.
+
+
+**Downsides**:
+
+This requires waiting on other features which might take a signficant time to land. It also could end up not fully fixing the problem (e.g. pushing down port-forwarding to the runtime doesn't ensure the runtime doesn't rely on host utilities).
+
+The Flex Volume feature is several releases out from fully replacing the current volumes as well.
+
+Finally, it's likely there are dependencies that neither of these proposals cover.
+
+## Non-Alternatives
+
+#### Pod + containerized flag
+
+Currently, there's a `--containerized` flag. This flag doesn't actually remove the dependency on mount utilities on the node though, so does not solve the problem described here. It also is under consideration for [removal](https://issues.k8s.io/18776).
+
+## Open Questions
+
+#### Why not a mount namespace?
+
+#### Timeframe
+
+1.6?
```

Comment:
We can get started with testing right away. We can decide on a release based on the testing results.